Enable Node Autocompletion for output ports of Dynamo nodes.

reddyashish opened the pull request:
### Purpose

This PR is to address the task: https://jira.autodesk.com/browse/DYN-3344

This PR will enable the node autocompletion functionality on the output ports of the nodes. The node autocompletion model will find all the matching nodes which has atleast one input parameter matching the output port type of the selected node. When any of the search elements are clicked, it will create the node and connect the output port of the original node to the first available matching input port of the new node. 

![autocompletion output ports](https://user-images.githubusercontent.com/43763136/115681882-f1731a00-a322-11eb-81e3-f460ac9a9a04.gif)


**Things to discuss**
1) When no matching search elements are found, do we want to return an empty list? Because the default nodes shown for the input nodes do not have any InPorts so we cannot show them for the output port. 

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

@QilongTang @zeusongit @mjkkirschner @aparajit-pratap

mjkkirschner commented:
@reddyashish maybe when nothing matches we show a list with codeblock, watch node, or python?

mjkkirschner commented on src/DynamoCore/Graph/Nodes/PortModel.cs:
```diff
@@ -440,6 +440,40 @@ internal string GetInputPortType()
+        {
+            if (PortType == PortType.Input) return null;
+
+            var ztNode = Owner as DSFunction;
```
instead of doing type checking like this using as and var you can probably switch to:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/is#pattern-matching-with-is

maybe a little cleaner.

QilongTang replied:
Ah I can see you are following the same approach as `GetInPortType`, do you mind correct both in your PR?
```
 If (ztNode is DSFunction)
```

reddyashish replied:
Changed this to use the "is" operator

mjkkirschner commented on src/DynamoCore/Graph/Nodes/PortModel.cs:
```diff
@@ -440,6 +440,40 @@ internal string GetInputPortType()
+
+                var outPortAttribute = classType.GetCustomAttributes().OfType<OutPortTypesAttribute>().FirstOrDefault();
+
+                return outPortAttribute?.PortTypes.ElementAt(Index);
```
you can't guarantee that just because you have the `outputAtrribute` that your index won't be out of bounds. I think `ElementAt` might throw - you may want to bounds check first, or use a try catch.

QilongTang replied:
yeah, 3rd party dev or even us could forgot to match the number of `OutPortTypesAttribute`, I have fixed one instance in my other PR

reddyashish replied:
Added a try-catch block.

mjkkirschner commented on src/DynamoCore/Search/SearchElements/NodeSearchElement.cs:
```diff
@@ -29,6 +29,7 @@ private void OnVisibilityChanged()
             if (VisibilityChanged != null)
                 VisibilityChanged();
         }
+        internal int PortToConnect { get; set; }
```
this seems like an inappropriate place for this data

QilongTang replied:
I kinda agree, if needed, we may want to inherit this class. I think in the future we may add more properties if need, e.g. favorite, weight, confidence level (after ML kicks in).

reddyashish replied:
Since this data will be specific to each search element in the auto-completion window, do we want to create a new class for these node elements and that would inherit the NodeSearchElement class?

QilongTang replied:
@reddyashish Yeah, I would be supportive

QilongTang commented on src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs:
```diff
@@ -418,7 +418,7 @@ private bool CanAutoComplete(object parameter)
             DynamoViewModel dynamoViewModel = _node.DynamoViewModel;
             // If the feature is enabled from Dynamo experiment setting and if user interaction is on input port.
-            return dynamoViewModel.EnableNodeAutoComplete && this.PortType == PortType.Input;
+            return dynamoViewModel.EnableNodeAutoComplete;
```
👍🏻

QilongTang commented:
> @reddyashish maybe when nothing matches we show a list with codeblock, watch node, or python?

Due to the convenience of creating CBN, I would maybe just add watch, watch 3D and Python, even watch3D may not be universal. My 2 cents.

reddyashish commented:
@mjkkirschner @QilongTang Addressed the comments. 
For the output ports, I am modified it to only show the watch, watch 3D and Python nodes as the CBN node does not have any input parameters.
To store the data for Auto-completion node elements, I created a new class where we can add more attributes in the future. The NodeSearchElement is extending this class as the auto-completion API returns a list of NodeSearchElement and we would need this information in the NodeSearchElementViewModel to connect the port. What do you think about it?

QilongTang commented on src/DynamoCore/Search/SearchElements/AutoCompletionNodeSearchElement.cs:
```diff
@@ -0,0 +1,16 @@
+    /// <summary>
+    ///      Class for all Auto-Completion Node search elements.
+    /// </summary>
+    public abstract class AutoCompletionNodeSearchElement
```
Can you re-iterate the purpose of this class by comments here? We cant inherit the NodeSearchElement class because auto-completion API returns a list of NodeSearchElement and we would need this information in the NodeSearchElementViewModel to connect the port? I am fine with this, @mjkkirschner @aparajit-pratap what do you think? I think we plan to keep expanding here because we will have more properties coming in specific to AutoComplete function, e.g. user's rank, whether favored, ML rank
